### PR TITLE
feat: add Polishing and Summary Features

### LINKED ISSUE #588: 🚀 功能建议：Easydict 支持「润色/语法检查+总结」
### 请先确认以下事项

- [X] 已仔细阅读了 [README](https://github.com/tisfeng/Easydict#readme)
- [X] 在 [issues](https://github.com/tisfeng/Easydict/issues) 页面搜索过（包括已关闭的 issue），未发现类似功能建议
- [X] Easydict 已升级到 [最新版本](https://github.com/tisfeng/Easydict/releases)
- [X] 我理解并认可上述内容，并理解项目维护者精力有限，**不遵循规则的 issue 可能会被无视或直接关闭**

### 功能描述

类似于 openai-translator，多加两个按钮用于 polishing 和 summary。
image

### 使用场景

Easydict 使用快捷键直接唤出可应用于一切文本：
- 替代 Grammarly 逐个词报错，更好优化需要开会员，且需要单独下软件，不能应用于全局
- 替代 LanguageTool 这种插件和桌面软件都要下的软件，且 Safari 支持不太好，软件适配不够多（读取文本是否安全？）
- 不需要老是打开 AI 客户端复制粘贴去润色/总结句子

### 实现方案（可选）

由于本人不会 swift 无法贡献代码，但可以给出一些拙见：
1. 关于 prompt: 看了#260，可以征集好的 prompt 写进代码里，不用实现自定义 prompt 对开发者友好
2. 关于选择模型：可以直接使用已经配置好的模型？或者重新填写亦可

或者可以参考 openai-translator 的关于这方面的实现 [PR](https://github.com/openai-translator/openai-translator/pull/719)

### 是否愿意提交 PR 实现该功能

- [ ] 我愿意提交 PR

### PR DESCRIPTION
**Pull request by Jerry23011**
closes #588 
截屏2024-07-12 21 50 31

### CODE REVIEW

File: Easydict/Swift/Service/LLMDerivs/SummaryService.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        // Set EZQueryTextType as translation since no formatting is needed
+        let queryType = EZQueryTextType.translation
+        let chatParam = LLMDerivParam(
+            text: text,
+            sourceLanguage: from,
+            serviceType: .summary
+        )
+
+        let chatHistory = serviceChatMessage(chatParam)
```
Comment (tisfeng) [inline Easydict/Swift/Service/LLMDerivs/SummaryService.swift:47]: "`SummaryService` and `PolishingService` have too much duplicate code; we can extract the prompt  code, add a new method called `chatHistory` to `BaseOpenAIService`, and they would only need to override this method.

Comment (tisfeng) [reply Easydict/Swift/Service/LLMDerivs/SummaryService.swift:47]: Sorry, I forgot there's already a method `serviceChatMessageModels()`, so `SummaryService` and `PolishingService` just need to override it.

Comment (tisfeng) [reply Easydict/Swift/Service/LLMDerivs/SummaryService.swift]: I've refactored the code, now we just need to simply rewrite the `chatMessageDicts()` method.

Comment (Jerry23011) [reply Easydict/Swift/Service/LLMDerivs/SummaryService.swift]: Which part of `chatMessageDicts()` do we need to rewrite and what do we want to achieve?

Comment (tisfeng) [reply Easydict/Swift/Service/LLMDerivs/SummaryService.swift]: For custom AI tool service, you only need to rewrite `chatMessageDicts` to return a prompt in the format of `[[String: String]]`. 

In the `serviceChatMessageModels` method of `BaseOpenAIService`, we will automatically convert the dictionary format into the `ChatMessage` format required by `OpenAI`.

Please refer to the code for details.

image

**Comment by tisfeng** [comment]
I tried `Summary` and it didn't work very well.

First, it should answer in the user's first preferred language, e.g. if I set Chinese as my first preferred language, it should summarize in Chinese at this point.

Second, the summary is too long, the summary should be concise and clear, the prompt may need to be improved.

This is just an experiment to test the AI tool service,  if `Summary` is not good, it's ok to only add `Polishing`.

**Comment by Jerry23011** [comment]
> First, it should answer in the user's first preferred language

That's very true, I added `Your summary should also be in \(sourceLanguage.queryLanguageName).` in the prompts to explicitly tell the model to output the `queryLanguage`

> Second, the summary is too long, the summary should be concise and clear, the prompt may need to be improved.

I added some examples for Chinese, hope it'll improve on performance for Chinese summary
截屏2024-07-13 18 57 25

**Comment by tisfeng** [comment]
> First, it should answer in the user's first preferred language, e.g. if I set Chinese as my first preferred language, it should summarize in Chinese at this point.

No, what I mean is, even if the source language of the text here is English, it should still be summarized in Chinese(first preferred language).

**Comment by Jerry23011** [comment]
> Even if the source language of the text here is English, it should still be summarized in Chinese

Can I add a toggle in settings to let users choose if they want to use the preferred language for the summary? I would expect for the summary to be the same as the source language.

**Comment by tisfeng** [comment]
I think it would be better to answer these queries in the user's first preferred language, so that it is easier for users to look up any language they are not familiar with, not just Chinese or English。

For example I could use it to summarize articles in Ukrainian.

The same logic is currently used for `Sentence` and `Dictionary`,  it would be better to keep it consistent.

**Comment by tisfeng** [comment]
For these unconventional AI tool services, it's better to set their usage status to `alwaysOff` by default, and only use them manually when needed 91ca4c07

**Comment by tisfeng** [comment]
For AI services, prompts are very important. 

These prompts may need further optimization later to achieve optimal results.

**Review by tisfeng — APPROVED** [review]
Looks good.

**Comment by Jerry23011** [comment]
> For AI services, prompts are very important.
> 
> These prompts may need further optimization later to achieve optimal results.

I’ll look it up later and see how we can improve them

File: Easydict/App/Localizable.xcstrings [hunk 2]
```diff
@@ -3834,6 +3850,16 @@
         }
       }
     },
+    "summary_service" : {
+      "localizations" : {
+        "zh-Hans" : {
+          "stringUnit" : {
+            "state" : "translated",
+            "value" : "总结"
```
Comment (phlpsong) [inline Easydict/App/Localizable.xcstrings:3858]: Looks like this missing the `en` localization.

Screenshot 2024-07-14 at 11 32 29

Comment (Jerry23011) [reply Easydict/App/Localizable.xcstrings:3858]: My bad, thank you

**Review by phlpsong — DISMISSED** [review]

File: Easydict/App/Localizable.xcstrings (hunk 2 above)
Comment (Jerry23011) [reply Easydict/App/Localizable.xcstrings:3858]: Sorry I didn’t take my laptop with me. Please let me fix it tonight

Comment (tisfeng) [reply Easydict/App/Localizable.xcstrings:3864]: I have added this en localization.

**Comment by tisfeng** [comment]
The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.

File: Easydict/App/Localizable.xcstrings (hunk 2 above)
Comment (phlpsong) [reply Easydict/App/Localizable.xcstrings:3864]: OK, it's my bad, I thought it has been fixed and not check the commit. 🥲

**Comment by Jerry23011** [comment]
> The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.
> 
> 

Yes indeed, thanks for pointing out. Let me check it tonight

**Comment by Jerry23011** [comment]
> The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.

Fixed with https://github.com/tisfeng/Easydict/pull/609/commits/72ccb48f45a8ca2006fe6745d0df9b13f3742e3f

**Comment by tisfeng** [comment]
ok, polishing icon looks good now, but summary icon still looks unclear, are there any other alternatives?

**Comment by Jerry23011** [comment]
They’re generated by DALL·E, let me try some new ones tomorrow

**Comment by Jerry23011** [comment]
I've added a new icon
截屏2024-07-16 14 51 28

**Comment by tisfeng** [comment]
ok, looks good now.

**Review by tisfeng — APPROVED** [review]

**Pull request merged**